### PR TITLE
[InstCombine] Improve folding of `icmp pred (and X, Mask/~Mask), Y)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add tests for expanding `foldICmpWithLowBitMaskedVal`; NFC
- [InstCombine] Improve mask detection in `foldICmpWithLowBitMaskedVal`
- [InstCombine] Recognize `(icmp eq/ne (and X, ~Mask), 0)` pattern in `foldICmpWithLowBitMaskedVal`

Proofs for new patterns: https://alive2.llvm.org/ce/z/fSQ3nZ

### CODE REVIEW

**Review by AZero13 — APPROVED** [review]
These changes look great!

File: llvm/include/llvm/IR/PatternMatch.h [hunk 1]
```diff
@@ -564,6 +564,19 @@ inline api_pred_ty<is_negated_power2> m_NegatedPower2(const APInt *&V) {
   return V;
 }
 
+struct is_negated_power2_or_zero {
```
Comment (dtcxzyw) [inline llvm/include/llvm/IR/PatternMatch.h:567]: Could you add some unit tests in `llvm/unittests/IR/PatternMatch.cpp`?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -4068,11 +4068,115 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
   return nullptr;
 }
 
+// Returns of V is a Mask ((X + 1) & X == 0) or ~Mask (-Pow2OrZero)
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4071]:
```suggestion
// Returns whether V is a Mask ((X + 1) & X == 0) or ~Mask (-Pow2OrZero)
```

File: llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll [hunk 3]
```diff
@@ -0,0 +1,629 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+declare void @llvm.assume(i1)
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll:4]: @nikic Do we still need to declare these intrinsic functions explicitly?

See also https://github.com/llvm/llvm-project/pull/78251.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 4]
```diff
@@ -4068,11 +4068,115 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+                         unsigned Depth = 0) {
+  if (Not ? match(V, m_NegatedPower2OrZero()) : match(V, m_LowBitMaskOrZero()))
+    return true;
+  if (V->getType()->getScalarSizeInBits() == 1)
+    return true;
+  if (Depth++ >= MaxAnalysisRecursionDepth)
+    return false;
+  Value *X;
+  if (match(V, m_Not(m_Value(X))))
+    return isMaskOrZero(X, !Not, Q, Depth);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4082]: Why not move the logic into the following switch statement?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 5]
```diff
@@ -4068,11 +4068,115 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+  const Operator *I = dyn_cast<Operator>(V);
+  if (I == nullptr)
+    return false;
+  switch (I->getOpcode()) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4086]: Could you please provide the alive2 proofs?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 6]
```diff
@@ -4068,11 +4068,115 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+  case Instruction::And:
+  case Instruction::Or:
+    // Mask0 | Mask1 is a Mask.
+    // Mask0 & Mask1 is a Mask.
+    // ~Mask0 | ~Mask1 is a ~Mask.
+    // ~Mask0 & ~Mask1 is a ~Mask.
+    return isMaskOrZero(I->getOperand(1), Not, Q, Depth) &&
+           isMaskOrZero(I->getOperand(0), Not, Q, Depth);
+  case Instruction::Xor:
+    // (X ^ (X - 1)) is a Mask
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4103]: Is it a Mask or a ~Mask?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4103]: Mask: its the [blsmsk](https://www.felixcloutier.com/x86/blsmsk) pattern from BMI2 extension.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 5 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4086]: Sorry, in the commit no the summary: https://alive2.llvm.org/ce/z/fSQ3nZ
added to summary as well.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 7]
```diff
@@ -4068,11 +4068,115 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+    return match(V, m_c_Xor(m_Value(X), m_Add(m_Deferred(X), m_AllOnes())));
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4104]:
```suggestion
    return  !Not && match(V, m_c_Xor(m_Value(X), m_Add(m_Deferred(X), m_AllOnes())));
```

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4104]: Ahh, good catch. Adding a test.

File: llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll (hunk 3 above)
Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll:4]: Don't seem to need it. Dropping in v2

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 8]
```diff
@@ -4068,11 +4068,117 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
 }
 
+// Returns whether V is a Mask ((X + 1) & X == 0) or ~Mask (-Pow2OrZero)
+static bool isMaskOrZero(const Value *V, bool Not, const SimplifyQuery &Q,
+                         unsigned Depth = 0) {
+  if (Not ? match(V, m_NegatedPower2OrZero()) : match(V, m_LowBitMaskOrZero()))
+    return true;
+  if (V->getType()->getScalarSizeInBits() == 1)
+    return true;
+  if (Depth++ >= MaxAnalysisRecursionDepth)
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4078]: What is the compile-time impact of this patch?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 9]
```diff
@@ -4068,11 +4068,117 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+    // (X ^ (X - 1)) is a Mask
+    return !Not &&
+           match(V, m_c_Xor(m_Value(X), m_Add(m_Deferred(X), m_AllOnes())));
+  case Instruction::Select:
+    // c ? Mask0 : Mask1 is a Mask.
+    return isMaskOrZero(I->getOperand(1), Not, Q, Depth) &&
+           isMaskOrZero(I->getOperand(2), Not, Q, Depth);
+  case Instruction::Shl:
+    if (Not) {
+      // (-1 >> X) << X is ~Mask
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4113]: Adding a patch. Will post seperately shortly.

[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4113]: see: https://github.com/llvm/llvm-project/pull/81576

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 8 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4078]: No meaningful affect: https://llvm-compile-time-tracker.com/compare.php?from=fb48fd18c240574841378defacadff34238089bb&to=ec40daf22da38fb28165ea1e757493470fec218e&stat=instructions:u

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4113]: Ill post v3 after this gets in.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 10]
```diff
@@ -4163,7 +4293,9 @@ static Value *foldICmpWithLowBitMaskedVal(ICmpInst &I,
     M = Constant::replaceUndefsWith(VecC, SafeReplacementConstant);
   }
 
-  return Builder.CreateICmp(DstPred, X, M);
+  if (NeedsNot)
+    M = IC.Builder.CreateNot(M);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4297]: Use `IC.getFreelyInverted`?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4297]: If its freely invertable `(not M)` will simplify later on. Seems simpler to let the existing pipeline clean it up. No?

**Comment by goldsteinn** [comment]
ping

**Comment by goldsteinn** [comment]
ping

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Comment by goldsteinn** [comment]
rebased

**Comment by goldsteinn** [comment]
ping @nikic

**Comment by goldsteinn** [comment]
ping

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 11]
```diff
@@ -4081,21 +4171,45 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+      if (IC.isFreeToInvert(X, X->hasOneUse()) && CheckMask(X, /*Not*/ true)) {
+        std::swap(X, M);
+        return true;
+      }
+      return IC.isFreeToInvert(M, M->hasOneUse()) && CheckMask(M, /*Not*/ true);
+    }
+    return false;
+  };
 
+  if (!TryMatch(0) && !TryMatch(1))
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4211]: Can you instead call this fold from foldICmpCommutative() and only handle one direction here?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 12]
```diff
@@ -4068,11 +4068,101 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+                         unsigned Depth = 0) {
+  if (Not ? match(V, m_NegatedPower2OrZero()) : match(V, m_LowBitMaskOrZero()))
+    return true;
+  if (V->getType()->getScalarSizeInBits() == 1)
+    return true;
+  if (Depth++ >= MaxAnalysisRecursionDepth)
+    return false;
+  Value *X;
+  const Operator *I = dyn_cast<Operator>(V);
+  if (I == nullptr)
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4082]: I dont think its an issue for llvm, but on systems where `nullptr` is non-zero, does `!I` still work?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 11 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4211]: Done, commit to do so is split into NFC.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 13]
```diff
@@ -4082,23 +4172,37 @@ Instruction *InstCombinerImpl::foldSelectICmp(ICmpInst::Predicate Pred,
...
+                                          InstCombiner &IC) {
+
+  Value *X, *M;
+  bool NeedsNot = false;
+
+  auto CheckMask = [&](Value *V, bool Not) {
+    if (!ICmpInst::isSigned(Pred))
+      return isMaskOrZero(V, Not, Q);
+    return Not ? match(V, m_NegatedPower2OrZero())
+               : match(V, m_LowBitMaskOrZero());
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4186]: They seem about the same to me, happy to change.

**Pull request closed without merging**